File: scripts/combine_sft.py
#!/usr/bin/env python3
"""Bundle rendered SFT JSONL files with duplicate-ID and template checks."""
from __future__ import annotations

import argparse
import hashlib
import json
import os
from pathlib import Path


RENDERER_FIELDS = ("template_sha256", "end_token", "terminal_tool_policy",
                   "teacher_reasoning_policy")


def combine(destination: Path, sources: list[Path]) -> dict:
    if destination.exists():
        raise ValueError(f"refusing overwrite: {destination}")
    if not sources or len({p.resolve() for p in sources}) != len(sources):
        raise ValueError("supply distinct source files")
    destination.parent.mkdir(parents=True, exist_ok=True)
    stage = destination.with_suffix(destination.suffix + f".building.{os.getpid()}")
    identities = set()
    renderer = None
    inputs = []
    pairs = reasoning_pairs = 0
    output_hash = hashlib.sha256()
    try:
        with stage.open("xb") as target:
            for path in sources:
                manifest_path = path.with_suffix(path.suffix + ".manifest.json")
                manifest = json.loads(manifest_path.read_text()) if manifest_path.exists() else None
                identity = ({key: manifest["renderer"].get(key) for key in RENDERER_FIELDS}
                            if manifest and manifest.get("renderer") else None)
                if identity is not None:
                    if renderer is not None and identity != renderer:
                        raise ValueError(f"incompatible SFT renderer: {path}")
                    renderer = identity
                source_hash = hashlib.sha256()
                count = 0
                with path.open("rb") as stream:
                    for number, line in enumerate(stream, 1):
                        source_hash.update(line)
                        if not line.endswith(b"\n"):
                            raise ValueError(f"{path}:{number}: missing JSONL newline")
                        row = json.loads(line)
                        if row["id"] in identities:
                            raise ValueError(f"duplicate SFT id: {row['id']}")
                        identities.add(row["id"])
                        target.write(line)
                        output_hash.update(line)
                        count += 1
                        reasoning_pairs += "<think>" in row["completion"]
                inputs.append({"path": str(path), "sha256": source_hash.hexdigest(),
                               "pairs": count, "manifest": str(manifest_path) if manifest else None})
                pairs += count
        stage.replace(destination)
    finally:
        stage.unlink(missing_ok=True)
    result = {"schema": "natlang.sft_bundle/1", "sources": inputs,
              "renderer": renderer, "pairs": pairs, "reasoning_pairs": reasoning_pairs,
              "sha256": output_hash.hexdigest()}
    destination.with_suffix(destination.suffix + ".manifest.json").write_text(
        json.dumps(result, indent=2) + "\n")
    return result


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("destination", type=Path)
    parser.add_argument("sources", nargs="+", type=Path)
    args = parser.parse_args()
    result = combine(args.destination, args.sources)
    print(f"{result['pairs']} pairs, {result['reasoning_pairs']} with reasoning -> {args.destination}")


if __name__ == "__main__":
    main()

File: ts-host/scripts/combine-sft.mjs
#!/usr/bin/env node
import { createHash, randomUUID } from 'node:crypto';
import { mkdir, readFile, rename, unlink, writeFile } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const RENDERER_FIELDS = ['template_sha256', 'end_token', 'terminal_tool_policy',
  'tokenizer_fingerprint_sha256', 'local_tokenizer_artifacts_sha256',
  'teacher_reasoning_policy', 'cache_stable_tools', 'native_roundtrip',
  'native_target_policy', 'invalid_action_policy'];
const sha256 = value => createHash('sha256').update(value).digest('hex');
const canonical = value => JSON.stringify(value, Object.keys(value ?? {}).sort());

async function optionalJson(path) { try { return JSON.parse(await readFile(path, 'utf8')); } catch (error) {
  if (error?.code === 'ENOENT') return null; throw error;
} }

export async function combineSft(destination, sources) {
  if (!sources.length || new Set(sources.map(source => resolve(source))).size !== sources.length)
    throw new Error('supply distinct source files');
  const identities = new Set(), inputs = [], chunks = []; let renderer = null, pairs = 0, reasoningPairs = 0;
  for (const source of sources) {
    const bytes = await readFile(source);
    if (bytes.length && bytes.at(-1) !== 10) throw new Error(`${source}: missing JSONL newline`);
    const manifestPath = `${source}.manifest.json`, manifest = await optionalJson(manifestPath);
    if (manifest?.renderer) {
      const identity = Object.fromEntries(RENDERER_FIELDS.map(key => [key, manifest.renderer[key] ?? null]));
      if (renderer && canonical(identity) !== canonical(renderer)) throw new Error(`incompatible SFT renderer: ${source}`);
      renderer = identity;
    }
    let count = 0;
    for (const [index, line] of bytes.toString('utf8').split('\n').entries()) {
      if (!line) continue;
      let row; try { row = JSON.parse(line); } catch (error) { throw new Error(`${source}:${index + 1}: ${error.message}`); }
      if (!row.id) throw new Error(`${source}:${index + 1}: missing id`);
      if (identities.has(row.id)) throw new Error(`duplicate SFT id: ${row.id}`);
      identities.add(row.id); count++; pairs++;
      if (String(row.completion ?? '').includes('<think>')) reasoningPairs++;
    }
    inputs.push({ path: source, sha256: sha256(bytes), pairs: count, manifest: manifest ? manifestPath : null });
    chunks.push(bytes);
  }
  const data = Buffer.concat(chunks), target = resolve(destination), staged = `${target}.building-${process.pid}-${randomUUID()}`;
  await mkdir(dirname(target), { recursive: true });
  await writeFile(staged, data, { flag: 'wx' });
  try { await rename(staged, target); } catch (error) { await unlink(staged).catch(() => {}); throw error; }
  const result = { schema: 'natlang.sft_bundle/1', sources: inputs, renderer,
    pairs, reasoning_pairs: reasoningPairs, sha256: sha256(data) };
  await writeFile(`${target}.manifest.json`, JSON.stringify(result, null, 2) + '\n', { flag: 'wx' });
  return result;
}

async function main() {
  const [destination, ...sources] = process.argv.slice(2);
  if (!destination || !sources.length) throw new Error('usage: combine-sft.mjs DESTINATION SOURCE...');
  const result = await combineSft(destination, sources);
  console.log(`${result.pairs} pairs, ${result.reasoning_pairs} with reasoning -> ${destination}`);
}

if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url))
  main().catch(error => { console.error(error instanceof Error ? error.message : String(error)); process.exitCode = 1; });
